### PR TITLE
[Check][Ability] Add Adaptability Ability Checkup 

### LINKED ISSUE #764: [Ability][Checkup] Adaptability
## Ability Requirements
- [x] Documentation
- [x] Tests
- [x] Adaptability increases STAB of a Pokémon with this Ability from 1.5 to 2.
- [x] When a Pokémon with Adaptability is Terastallized, Adaptability only affects moves of the Tera Type. If the Pokémon's Tera Type is one of its original types, its STAB is 2.25; otherwise, it's 2.
- [x] Stellar-type moves are unaffected by Adaptability. 

## Related Bugs
- Adaptability-Terastalization interaction is incorrect

## References
- [Bulbapedia - Adaptability](https://bulbapedia.bulbagarden.net/wiki/Adaptability_(Ability))

### LINKED ISSUE #792: [Bug] Adaptability is Activated by Struggle
### Describe the bug

Struggle is affected by Adaptability

Link to Discord's Bug Report (Adaptability trigger on normal type struggle)
https://discord.com/channels/1125469663833370665/1347317663650807859

### Reproduction

Use Rapid Spin until the opponent faints, then use Struggle.

```
	STARTER_SPECIES_OVERRIDE: Species.RATTATA,
	MOVESET_OVERRIDE: [Moves.RAPID_SPIN],
	ABILITY_OVERRIDE: Abilities.ADAPTABILITY,

	OPP_LEVEL_OVERRIDE: 3,
	OPP_SPECIES_OVERRIDE: Species.SHROOMISH,
	OPP_MOVESET_OVERRIDE: [Moves.GRUDGE],
```

### Expected behavior

https://replay.pokemonshowdown.com/gen9customgame-2314011640

Struggle should be unaffected by Adaptability.

### Screenshots / Videos

https://github.com/user-attachments/assets/4cd6fbbd-df89-46e2-add2-b0a5b1080306

### Session export file

_No response_

### User data export file

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by flx-sta**
## What are the changes the user will see?

[Adaptability](https://bulbapedia.bulbagarden.net/wiki/Adaptability_(Ability)) will no longer increase the [STAB](https://bulbapedia.bulbagarden.net/wiki/Same-type_attack_bonus) (by 0.5) on the users moves if tera-type doesn't match users original types

## Why am I making these changes?

Covering the abilities behavior + edge cases with automated tests.
Fixing the ability currently applying to tera-type AND original type (if they don't match)

### Stab Table

> Using fire FIRE and water WATER as example types

| User Type | Tera Type | Move Type | Expected STAB | w/ Adaptability |
|-----------|-----------|-----------|---------------|-----------------|
|  | - |  | 1.0 | 1.0 |
|  | - |  | 1.5 | 2.0 |
|  | - | PLEDGE COMBO | 1.5 | 2.0 |
|  |  |  | 1.5 | 1.5 |
|  |  |  | 1.5 | 2.0 |
|  |  | PLEDGE COMBO | 1.5 | 2.0 |
|  |  |  | 1.0 | 1.0 |
|  |  |  | 2.0 | 2.25 |
|  |  | PLEDGE COMBO | 2.0 | 2.25 |
|  |  |  | 1.5 | 1.5 |

<sup>Images &copy; Bulbapedia</sup>

## What are the changes from a developer perspective?

- Add multiple default stab tests
- Add Adaptability related tests (being STAB calculations)
- Move stab calculation into it's own, public, method inside the `Pokemon` class
- Fix a bug that STAB was increased by 0.5 even if tera-type didn't match users original type

## How to test the changes?

Run the tests

## Checklist

- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`npm run test:silent`)
  - [x] Have I created new automated tests (`npm run test:create`) or updated existing tests related to the PR's changes?

### CODE REVIEW

**Comment by DayKev** [comment]
> | User Type | Tera Type | Move Type | Expected STAB w/ Adaptability |
> |--------|--------|--------|--------|
> | fire FIRE | - | water WATER | 1.0 |
> | fire FIRE | - | fire FIRE | 2.0 |
> | fire FIRE | water WATER | fire FIRE | 1.5 |
> | fire FIRE | water WATER | water WATER | 1.5 | 
> | fire FIRE | fire FIRE | fire FIRE | 2.25 | 

Adaptability is supposed to boost the user's original types if the user is not Terastallized, and only the user's Tera type if the user is Terastallized. So Tera Water Charmander using Water Gun while Terastallized should get a 2x STAB boost.

**Comment by flx-sta** [comment]
> > [table]
> 
> Adaptability is supposed to boost the user's original types if the user is not Terastallized, and only the user's Tera type if the user is Terastallized. So Tera Water Charmander using Water Gun while Terastallized should get a 2x STAB boost.

Fixed. Thanks

File: src/data/abilities/ab-attrs/stab-boost-ab-attr.ts [hunk 1]
```diff
@@ -1,18 +1,45 @@
...
-      return true;
+  override apply(pokemon: Pokemon, _simulated: boolean, move: Move, stabMultiplier: NumberHolder): boolean {
+    if (pokemon.getTypes().includes(move.type)) {
+      if (pokemon.isTerastallized()) {
+        const moveType = move.type;
+        const teraType = pokemon.getTeraType();
+
+        if (moveType === teraType) {
+          stabMultiplier.value += 0.5; // Adaptability only applies if the move type is the same as the tera type (When the pokemon is terastallized)
+          return true;
```
Comment (Tempo-anon) [inline src/data/abilities/ab-attrs/stab-boost-ab-attr.ts]: Would it make sense to also cap the STAB boost to 2.25 here instead of in `pokemon.ts`?
```ts
/** A damage multiplier for when the attack is of the attacker's type and/or Tera type. */
    const stabMultiplier = new NumberHolder(1);
    if (matchesSourceType) {
      stabMultiplier.value += 0.5;
    }
    applyMoveAttrs(CombinedPledgeStabBoostAttr, source, this, move, stabMultiplier);
    if (sourceTeraType !== ElementalType.UNKNOWN && sourceTeraType === moveType) {
      stabMultiplier.value += 0.5;
    }

    applyAbFunc<StabBoostAbAttr>(AbAttrFlag.STAB_BOOST, source, simulated, stabMultiplier);

    stabMultiplier.value = Math.min(stabMultiplier.value, 2.25);
```

File: test/abilities/adaptability.test.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    game.override
+      .battleType("single")
+      .startingLevel(1)
+      .moveset([MoveId.EMBER, MoveId.WATER_GUN])
+      .ability(AbilityId.ADAPTABILITY)
+      .enemySpecies(SpeciesId.MUNCHLAX)
+      .disableCrits();
+  });
+
```
Comment (Tempo-anon) [inline test/abilities/adaptability.test.ts:34]: There should also be a test either here or in the pledge moves for interactions with adaptability

File: src/field/pokemon.ts [hunk 3]
```diff
@@ -3060,22 +3060,9 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
...
-    if (sourceTeraType !== ElementalType.UNKNOWN && sourceTeraType === moveType) {
-      stabMultiplier.value += 0.5;
-    }
-
-    applyAbFunc<StabBoostAbAttr>(AbAttrFlag.STAB_BOOST, source, simulated, stabMultiplier);
-
-    stabMultiplier.value = Math.min(stabMultiplier.value, 2.25);
+    /** A damage multiplier for when the attack is of the same type as the attacker type/teraType. */
+    const stabMultiplier: number = this.calcStabMultiplierForTakingDamage(source, move, abilityApplyMode, simulated);
+    console.debug("Stab multiplier: ", stabMultiplier);
```
Comment (DayKev) [inline src/field/pokemon.ts]:
```suggestion
```
Leftover testing log?

File: test/pokemon/stab.test.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  });
+
+  it("should have a NO STAB (1.0) on type mismatch", async () => {
+    await game.classicMode.startBattle([SpeciesId.CHARMANDER]);
+
+    const enemyPokemon = game.field.getEnemyPokemon();
+    vi.spyOn(enemyPokemon, "calcStabMultiplierForTakingDamage");
+
+    game.move.use(MoveId.WATER_GUN);
+    await game.move.selectEnemyMove(MoveId.SPLASH);
```
Comment (DayKev) [inline test/pokemon/stab.test.ts]:
```suggestion
```
You only need to choose the enemy's move if the enemy moveset override has more than one entry or you're not using it at all.

File: test/pokemon/stab.test.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  });
+
+  it("should have a STAB of 1.5 on type match", async () => {
+    await game.classicMode.startBattle([SpeciesId.CHARMANDER]);
+
+    const enemyPokemon = game.field.getEnemyPokemon();
+    vi.spyOn(enemyPokemon, "calcStabMultiplierForTakingDamage");
+
+    game.move.use(MoveId.EMBER);
+    await game.move.selectEnemyMove(MoveId.SPLASH);
```
Comment (DayKev) [inline test/pokemon/stab.test.ts]:
```suggestion
```

File: test/pokemon/stab.test.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+    const playerPokemon = game.field.getPlayerPokemon();
+    expect(playerPokemon.isTerastallized()).toBe(true);
+    expect(playerPokemon.getTeraType()).toBe(ElementalType.WATER);
+
+    const enemyPokemon = game.field.getEnemyPokemon();
+    vi.spyOn(enemyPokemon, "calcStabMultiplierForTakingDamage");
+
+    game.move.use(MoveId.WATER_GUN);
+    await game.move.selectEnemyMove(MoveId.SPLASH);
```
Comment (DayKev) [inline test/pokemon/stab.test.ts]:
```suggestion
```

File: src/data/abilities/ab-attrs/stab-boost-ab-attr.ts [hunk 7]
```diff
@@ -1,18 +1,45 @@
...
     super(showAbility, showAbilityInstant);
     this._flags.add(AbAttrFlag.STAB_BOOST);
   }
 
-  override apply(_pokemon: Pokemon, _simulated: boolean, stabMultiplier: NumberHolder): boolean {
-    if (stabMultiplier.value > 1) {
-      stabMultiplier.value += 0.5;
-      return true;
+  override apply(pokemon: Pokemon, _simulated: boolean, move: Move, stabMultiplier: NumberHolder): boolean {
+    if (pokemon.getTypes().includes(move.type)) {
```
Comment (DayKev) [inline src/data/abilities/ab-attrs/stab-boost-ab-attr.ts]:
```suggestion
    if (!pokemon.getTypes().includes(move.type)) {
      return false;
    }
```

File: src/data/abilities/ab-attrs/stab-boost-ab-attr.ts (hunk 1 above)
Comment (flx-sta) [reply src/data/abilities/ab-attrs/stab-boost-ab-attr.ts]: It would, yes.
I didn't change it because I didn't want to touch too much existing code but that seems like a logical change to me

File: src/field/pokemon.ts (hunk 3 above)
Comment (flx-sta) [reply src/field/pokemon.ts]: I thought it might be helpful and is removed in production anyway but I can also remove it

File: test/abilities/adaptability.test.ts (hunk 2 above)
Comment (flx-sta) [reply test/abilities/adaptability.test.ts:34]: What are those interactions? That's the first time I hear about this

Comment (flx-sta) [reply test/abilities/adaptability.test.ts:34]: I'm not fond of pledge moves. I'd ask you to maybe look into these tests. I've put some placeholder `it.todo` in place

**Comment by flx-sta** [comment]
I've decided **not** to add the STAB tests for pledge moves as I simply don't know those moves at all.

I did put some `todo` tests in place though

File: src/field/pokemon.ts (hunk 3 above)
Comment (DayKev) [reply src/field/pokemon.ts]: `.debug` logs are not removed in production, only `.log` logs.

**Comment by flx-sta** [comment]
Added stellar tera-type checks

File: src/data/abilities/ab-attrs/stab-boost-ab-attr.ts [hunk 8]
```diff
@@ -1,20 +1,56 @@
...
+ */
 export class StabBoostAbAttr extends AbAttr {
   constructor(showAbility: boolean = true, showAbilityInstant: boolean = false) {
     super(showAbility, showAbilityInstant);
     this._flags.add(AbAttrFlag.STAB_BOOST);
   }
 
-  override apply(_pokemon: Pokemon, _simulated: boolean, stabMultiplier: NumberHolder): boolean {
-    if (stabMultiplier.value > 1) {
+  override apply(pokemon: Pokemon, _simulated: boolean, move: Move, stabMultiplier: NumberHolder): boolean {
```
Comment (Tempo-anon) [inline src/data/abilities/ab-attrs/stab-boost-ab-attr.ts:34]: Just a 🤔 but since this is the last step of stab application, we could also do stabMultiplier.value = 2.0 instead of just adding to it.

**Review by Tempo-anon — DISMISSED** [review]

File: src/data/abilities/ab-attrs/stab-boost-ab-attr.ts (hunk 8 above)
Comment (flx-sta) [reply src/data/abilities/ab-attrs/stab-boost-ab-attr.ts:34]: I'd rather keep it the way it is since that represents better what is documented on bulbapedia as code

**Comment by PigeonBar** [comment]
> I've decided **not** to add the STAB tests for pledge moves as I simply don't know those moves at all.
> 
> I did put some `todo` tests in place though

<details><summary>According to Showdown, Adaptability does increase the STAB boost given by Pledge moves. (Turns 1-2 are with Adaptability and aimed at the left Registeel. Turns 3-4 are with Gastro Acid suppressing Adaptability and aimed at the right Registeel.)</summary>
<p>

20250417 Adaptability Pledge Moves

</p>
</details>

File: src/field/pokemon.ts [hunk 9]
```diff
@@ -3060,22 +3060,8 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
...
-    applyMoveAttrs(CombinedPledgeStabBoostAttr, source, this, move, stabMultiplier);
-    if (sourceTeraType !== ElementalType.UNKNOWN && sourceTeraType === moveType) {
-      stabMultiplier.value += 0.5;
-    }
-
-    applyAbFunc<StabBoostAbAttr>(AbAttrFlag.STAB_BOOST, source, simulated, stabMultiplier);
-
-    stabMultiplier.value = Math.min(stabMultiplier.value, 2.25);
+    /** A damage multiplier for when the attack is of the same type as the attacker type/teraType. */
+    const stabMultiplier: number = this.calcStabMultiplierForTakingDamage(source, move, abilityApplyMode, simulated);
```
Comment (PigeonBar) [inline src/field/pokemon.ts]:
```suggestion
    const stabMultiplier = this.calcStabMultiplierForTakingDamage(source, move, abilityApplyMode, simulated);
```

File: src/data/abilities/ab-attrs/stab-boost-ab-attr.ts [hunk 10]
```diff
@@ -1,20 +1,56 @@
...
+import { AbAttr } from "#app/data/abilities/ab-attrs/ab-attr";
+import type { Move } from "#app/data/moves/move";
 import type { Pokemon } from "#app/field/pokemon";
 import type { NumberHolder } from "#app/utils";
 import { AbAttrFlag } from "#enums/ab-attr-flag";
-import { AbAttr } from "./ab-attr";
 
+/**
+ * Applies a STAB multiplier boost of `0.5` if the move type is the same as one of the pokemon's types
+ * unless the pokemon is terastallized. Then the STAB multiplier is only applied if the move type is the same as the tera type.
```
Comment (PigeonBar) [inline src/data/abilities/ab-attrs/stab-boost-ab-attr.ts]:
```suggestion
 * Applies a STAB multiplier boost of `0.5` if the move type is the same as one of the pokemon's types.
 * If the Pokemon is terastallized, then the STAB multiplier is only applied if the move type is the same as the tera type.
 * The total STAB multiplier is capped at 2.25.
 * 
```

File: src/data/abilities/ab-attrs/stab-boost-ab-attr.ts [hunk 11]
```diff
@@ -1,20 +1,56 @@
...
 export class StabBoostAbAttr extends AbAttr {
   constructor(showAbility: boolean = true, showAbilityInstant: boolean = false) {
     super(showAbility, showAbilityInstant);
     this._flags.add(AbAttrFlag.STAB_BOOST);
   }
 
-  override apply(_pokemon: Pokemon, _simulated: boolean, stabMultiplier: NumberHolder): boolean {
-    if (stabMultiplier.value > 1) {
+  override apply(pokemon: Pokemon, _simulated: boolean, move: Move, stabMultiplier: NumberHolder): boolean {
+    const initialStabMultiplier = stabMultiplier.value;
```
Comment (PigeonBar) [inline src/data/abilities/ab-attrs/stab-boost-ab-attr.ts:40]:
```suggestion
    // Adaptability does not apply to Stellar-type moves
    if (pokemon.getMoveType(move) === ElementalType.STELLAR) {
      return false;
    }

    const initialStabMultiplier = stabMultiplier.value;
```

File: src/data/abilities/ab-attrs/stab-boost-ab-attr.ts [hunk 12]
```diff
@@ -1,20 +1,56 @@
...
+
+    if (pokemon.isTerastallized()) {
+      if (pokemon.getTypes().includes(pokemon.getTeraType())) {
+        // If the tera type is one of the pokemon's original types then the STAB multiplier is increased by 0.25 (to 2.25)
+        stabMultiplier.value += 0.25;
+      } else if (pokemon.getTeraType() === move.type) {
```
Comment (PigeonBar) [inline src/data/abilities/ab-attrs/stab-boost-ab-attr.ts]:
```suggestion
      } else if (pokemon.getTeraType() === pokemon.getMoveType(move)) {
```

File: src/data/abilities/ab-attrs/stab-boost-ab-attr.ts [hunk 13]
```diff
@@ -1,20 +1,56 @@
...
+        // if the tera type is NOT one of the pokemon's original types but is the same as the move type then the STAB multiplier is increased by 0.5
+        stabMultiplier.value += 0.5;
+      }
+    } else if (pokemon.getTypes().includes(move.type)) {
```
Comment (PigeonBar) [inline src/data/abilities/ab-attrs/stab-boost-ab-attr.ts]:
```suggestion
    } else if (pokemon.getTypes().includes(pokemon.getMoveType(move))) {
```

File: test/abilities/adaptability.test.ts [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    await game.classicMode.startBattle([SpeciesId.CHARMANDER]);
+
+    const enemyPokemon = game.field.getEnemyPokemon();
+    vi.spyOn(enemyPokemon, "calcStabMultiplierForTakingDamage");
+
+    game.move.select(MoveId.EMBER);
+    await game.toEndOfTurn();
+
+    expect(enemyPokemon.calcStabMultiplierForTakingDamage).toHaveReturnedWith(2.0);
+  });
```
Comment (PigeonBar) [inline test/abilities/adaptability.test.ts:45]:
```suggestion
  });

  it("should increase STAB to 2.0 if move type changes to a type that matches one of the user's types", async () => {
    await game.classicMode.startBattle([SpeciesId.TYNAMO]);

    const enemyPokemon = game.field.getEnemyPokemon();
    vi.spyOn(enemyPokemon, "calcStabMultiplierForTakingDamage");

    game.setTurnOrder([BattlerIndex.ENEMY, BattlerIndex.PLAYER]);
    game.move.select(MoveId.EMBER);
    await game.move.forceEnemyMove(MoveId.ELECTRIFY);
    await game.toEndOfTurn();

    expect(enemyPokemon.calcStabMultiplierForTakingDamage).toHaveReturnedWith(2.0);
  });

  it("should not apply STAB if move type does not match one of the user's types", async () => {
    await game.classicMode.startBattle([SpeciesId.CHARMANDER]);

    const enemyPokemon = game.field.getEnemyPokemon();
    vi.spyOn(enemyPokemon, "calcStabMultiplierForTakingDamage");

    game.move.select(MoveId.WATER_GUN);
    await game.toEndOfTurn();

    expect(enemyPokemon.calcStabMultiplierForTakingDamage).toHaveReturnedWith(1.0);
  });
```

File: test/abilities/adaptability.test.ts [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+  describe("Terastallized", () => {
```
Comment (PigeonBar) [inline test/abilities/adaptability.test.ts:85]:
```suggestion
  describe("Terastallized", () => {
    it("should keep STAB at 1.5 if move type, but not tera type, is one of the user's original types", async () => {
      game.override.startingHeldItems([{ name: "TERA_SHARD", type: ElementalType.WATER }]);

      await game.classicMode.startBattle([SpeciesId.CHARMANDER]);

      const playerPokemon = game.field.getPlayerPokemon();

      expect(playerPokemon.isTerastallized()).toBe(true);
      expect(playerPokemon.getTeraType()).toBe(ElementalType.WATER);

      const enemyPokemon = game.field.getEnemyPokemon();
      vi.spyOn(enemyPokemon, "calcStabMultiplierForTakingDamage");

      game.move.select(MoveId.EMBER);
      await game.toEndOfTurn();

      expect(enemyPokemon.calcStabMultiplierForTakingDamage).toHaveReturnedWith(1.5);
    });

```

File: test/pokemon/stab.test.ts [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+    game = new GameManager(phaserGame);
+    game.override
+      .disableCrits()
+      .ability(AbilityId.BALL_FETCH)
+      .battleType("single")
+      .enemySpecies(SpeciesId.MAGIKARP)
+      .enemyMoveset(MoveId.SPLASH);
+  });
+
+  it("should have a NO STAB (1.0) on type mismatch", async () => {
```
Comment (DayKev) [inline test/pokemon/stab.test.ts]:
```suggestion
  it("should have NO STAB (1.0) on type mismatch", async () => {
```

File: src/data/abilities/ab-attrs/stab-boost-ab-attr.ts [hunk 17]
```diff
@@ -1,20 +1,57 @@
...
+  override apply(pokemon: Pokemon, _simulated: boolean, move: Move, stabMultiplier: NumberHolder): boolean {
+    // Adaptability does not apply to Stellar-type moves
+    if (pokemon.getMoveType(move) === ElementalType.STELLAR) {
+      return false;
+    }
+
+    const initialStabMultiplier = stabMultiplier.value;
+
+    if (pokemon.isTerastallized()) {
+      if (pokemon.getTypes().includes(pokemon.getTeraType())) {
```
Comment (PigeonBar) [inline src/data/abilities/ab-attrs/stab-boost-ab-attr.ts]:
```suggestion
      if (pokemon.getTypes().includes(pokemon.getTeraType()) && pokemon.getTeraType() === pokemon.getMoveType(move)) {
```

File: test/abilities/adaptability.test.ts [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+      expect(playerPokemon.getTeraType()).toBe(ElementalType.FIRE);
+
+      const enemyPokemon = game.field.getEnemyPokemon();
+      vi.spyOn(enemyPokemon, "calcStabMultiplierForTakingDamage");
+
+      game.move.select(MoveId.EMBER);
+      await game.toEndOfTurn();
+
+      expect(enemyPokemon.calcStabMultiplierForTakingDamage).toHaveReturnedWith(2.25);
+    });
```
Comment (PigeonBar) [inline test/abilities/adaptability.test.ts:141]:
```suggestion
    });

    it("should not apply STAB if move type does NOT match tera type or the user's original types", async () => {
      game.override.startingHeldItems([{ name: "TERA_SHARD", type: ElementalType.FIRE }]);

      await game.classicMode.startBattle([SpeciesId.CHARMANDER]);

      const playerPokemon = game.field.getPlayerPokemon();

      expect(playerPokemon.isTerastallized()).toBe(true);
      expect(playerPokemon.getTeraType()).toBe(ElementalType.FIRE);

      const enemyPokemon = game.field.getEnemyPokemon();
      vi.spyOn(enemyPokemon, "calcStabMultiplierForTakingDamage");

      game.move.select(MoveId.WATER_GUN);
      await game.toEndOfTurn();

      expect(enemyPokemon.calcStabMultiplierForTakingDamage).toHaveReturnedWith(1.0);
    });
```

File: src/field/pokemon.ts [hunk 19]
```diff
@@ -4285,6 +4271,40 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
...
+   * @param abilityApplyMode the {@linkcode AbilityApplyMode} determining how abilities are applied.
+   * @param simulated If `true`, suppresses changes to game state during the calculation.
+   * @returns A {@linkcode NumberHolder} containing the STAB multiplier as value
+   */
+  public calcStabMultiplierForTakingDamage(
+    source: Pokemon,
+    move: Move,
+    abilityApplyMode: AbilityApplyMode,
+    simulated: boolean,
+  ): number {
```
Comment (PigeonBar) [inline src/field/pokemon.ts:4288]:
```suggestion
  ): number {
    if (move.hasAttr(TypelessAttr)) {
      return 1;
    }

```

A band-aid fix for #792. (`TypelessAttr` doesn't actually set a move's type to `ElementalType.UNKNOWN`, so every part of the codebase treats Struggle as a Normal-type move, which is its own can of worms.)

File: test/abilities/adaptability.test.ts [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+    await game.classicMode.startBattle([SpeciesId.CHARMANDER]);
+
+    const enemyPokemon = game.field.getEnemyPokemon();
+    vi.spyOn(enemyPokemon, "calcStabMultiplierForTakingDamage");
+
+    game.move.select(MoveId.WATER_GUN);
+    await game.toEndOfTurn();
+
+    expect(enemyPokemon.calcStabMultiplierForTakingDamage).toHaveReturnedWith(1.0);
+  });
```
Comment (PigeonBar) [inline test/abilities/adaptability.test.ts:71]:
```suggestion
  });
  
  it("should not apply STAB to Struggle", async () => {
    await game.classicMode.startBattle([SpeciesId.RATTATA]);

    const enemyPokemon = game.field.getEnemyPokemon();
    vi.spyOn(enemyPokemon, "calcStabMultiplierForTakingDamage");

    game.move.use(MoveId.STRUGGLE);
    await game.toEndOfTurn();

    expect(enemyPokemon.calcStabMultiplierForTakingDamage).toHaveReturnedWith(1.0);
  });
```

**Comment by innerthunder** [comment]
#1014 to fix Pledge move interactions

**Review by innerthunder — DISMISSED** [review]

File: src/data/abilities/ab-attrs/stab-boost-ab-attr.ts [hunk 21]
```diff
@@ -1,20 +1,57 @@
...
+import { AbAttr } from "#app/data/abilities/ab-attrs/ab-attr";
+import type { Move } from "#app/data/moves/move";
 import type { Pokemon } from "#app/field/pokemon";
 import type { NumberHolder } from "#app/utils";
 import { AbAttrFlag } from "#enums/ab-attr-flag";
-import { AbAttr } from "./ab-attr";
+import { ElementalType } from "#enums/elemental-type";
 
+/**
+ * Applies a STAB multiplier boost of `0.5` if the move type is the same as one of the pokemon's types.
```
Comment (PigeonBar) [inline src/data/abilities/ab-attrs/stab-boost-ab-attr.ts]:
```suggestion
 * Increases the STAB multiplier by `+0.5` if the move type is the same as one of the pokemon's types.
```

**Review by PigeonBar — DISMISSED** [review]
